fdbserver/ptxn/test/RealTLog.actor.cpp

```diff
@@ -0,0 +1,274 @@
+	Reference<AsyncVar<bool>> isDegraded = FlowTransport::transport().getDegraded();
+	Reference<AsyncVar<UID>> activeSharedTLog(new AsyncVar<UID>(pTLogContext->tLogID));
+	state PromiseStream<InitializeTLogRequest> promiseStream = PromiseStream<InitializeTLogRequest>();
+	Promise<Void> oldLog;
+	Promise<Void> recovery;
+
+	// construct tLog.
+	state Future<Void> tl = tLog(pTLogContext->persistentData,
+	                             pTLogContext->persistentQueue,
+	                             dbInfoRef,
+	                             localities,
+	                             promiseStream,
+	                             pTLogContext->tLogID,
+	                             pTLogDriverContext->workerID,
+	                             false, /* restoreFromDisk */
+	                             oldLog,
+	                             recovery,
+	                             pTLogDriverContext->diskQueueBasename,
+	                             isDegraded,
+	                             activeSharedTLog);
+
+	// start tlog.
+	state InitializeTLogRequest initTlogReq = InitializeTLogRequest();
+	initTlogReq.isPrimary = true;
+	TLogInterface interface = wait(promiseStream.getReply(initTlogReq));
+	pTLogContext->realTLogInterface = interface;
+
+	// inform other actors tLog is ready.
+	pTLogContext->TLogCreated.send(true);
+
+	// wait for either test completion or abnormal failure.
+	choose {
+		when(wait(tl)) {}
+		when(bool testCompleted = wait(pTLogContext->realTLogTestCompleted.getFuture())) {}
+	}
+
+	return Void();
+}
+
+// sent commits through TLog interface.
+ACTOR Future<Void> getTLogCommitActor(std::shared_ptr<TestDriverContext> pTestDriverContext,
+                                      std::shared_ptr<TLogDriverContext> pTLogDriverContext) {
+	bool tLogReady = wait(pTLogDriverContext->pTLogContext->TLogStarted.getFuture());
+	state Version prev = 0;
+	state Version next = 1;
+	state int i = 0;
+	for (; i < pTLogDriverContext->numCommits; i++) {
+		Standalone<StringRef> key = StringRef(format("key %d", i));
+		Standalone<StringRef> val = StringRef(format("value %d", i));
+		MutationRef m(MutationRef::Type::SetValue, key, val);
+		Tag tag(pTLogDriverContext->tagLocality, pTLogDriverContext->pTLogContext->tagProcessID);
+
+		// build commit request
+		LogPushData toCommit(pTLogDriverContext->ls, pTLogDriverContext->pTLogContext->realTLogInterface);
+		UID spanID = deterministicRandom()->randomUniqueID();
+		toCommit.addTransactionInfo(spanID);
+		vector<Tag> tags = { tag };
+		toCommit.addTags(tags);
+		toCommit.writeTypedMessage(m);
+		int location = 0;
+		Standalone<StringRef> msg = toCommit.getMessages(location);
+
+		// send commit and wait for reply.
+		::TLogCommitRequest request(
+		    SpanID(), msg.arena(), prev, next, prev, prev, msg, deterministicRandom()->randomUniqueID());
+		::TLogCommitReply reply = wait(pTLogDriverContext->pTLogContext->realTLogInterface.commit.getReply(request));
+		prev++;
```

Comment:
Nit: `ASSERT` the `reply` message fields if possible.

Reply:
`reply.version` appears to be the durable version, which is updated asynchronously. I think the best we can assert here (unless we peek), is `reply.version<= next`